python/pyarrow/plasma.pyx

```diff
@@ -0,0 +1,398 @@
+
+    cdef cppclass CPlasmaClient" plasma::PlasmaClient":
+
+        CPlasmaClient()
+
+        CStatus Connect(const c_string& store_socket_name,
+                        const c_string& manager_socket_name, int release_delay)
+
+        CStatus Create(const CUniqueID& object_id, int64_t data_size,
+                       const uint8_t* metadata, int64_t metadata_size,
+                       uint8_t** data)
+
+        CStatus Get(const CUniqueID* object_ids, int64_t num_objects,
+                    int64_t timeout_ms, CObjectBuffer* object_buffers)
+
+        CStatus Seal(const CUniqueID& object_id)
+
+        CStatus Evict(int64_t num_bytes, int64_t& num_bytes_evicted)
+
+        CStatus Hash(const CUniqueID& object_id, uint8_t* digest)
+
+        CStatus Release(const CUniqueID& object_id)
+
+        CStatus Contains(const CUniqueID& object_id, c_bool* has_object)
+
+        CStatus Subscribe(int* fd)
+
+        CStatus GetNotification(int fd, CUniqueID* object_id,
+                                int64_t* data_size, int64_t* metadata_size)
+
+        CStatus Disconnect()
+
+
+cdef extern from "plasma/client.h" nogil:
+
+    cdef struct CObjectBuffer" plasma::ObjectBuffer":
+        int64_t data_size
+        uint8_t* data
+        int64_t metadata_size
+        uint8_t* metadata
+
+
+def make_object_id(object_id):
+    return ObjectID(object_id)
+
+
+cdef class ObjectID:
+
+    cdef:
+        CUniqueID data
+
+    def __cinit__(self, object_id):
+        self.data = CUniqueID.from_binary(object_id)
+
+    def __richcmp__(ObjectID self, ObjectID object_id, operation):
+        if operation != 2:
+            raise ValueError("operation != 2 (only equality is supported)")
+        return self.data == object_id.data
+
+    def __repr__(self):
+        return "ObjectID(" + self.data.hex().decode() + ")"
+
+    def __reduce__(self):
+        return (make_object_id, (self.data.binary(),))
+
+
+cdef class PlasmaBuffer(Buffer):
+    """This is the type returned by calls to get with a PlasmaClient.
+
+    We define our own class instead of directly returning a buffer object so
+    that we can add a custom destructor which notifies Plasma that the object
+    is no longer being used, so the memory in the Plasma store backing the
+    object can potentially be freed.
+
+    Attributes:
+        object_id (ObjectID): The ID of the object in the buffer.
+        client (PlasmaClient): The PlasmaClient that we use to communicate
+            with the store and manager.
+    """
+
+    cdef:
+        ObjectID object_id
+        PlasmaClient client
+
+    def __cinit__(self, ObjectID object_id, PlasmaClient client):
+        """Initialize a PlasmaBuffer."""
+        self.object_id = object_id
+        self.client = client
+
+    def __dealloc__(self):
+        """Notify Plasma that the object is no longer needed.
+
+        If the plasma client has been shut down, then don't do anything.
+        """
+        self.client.release(self.object_id)
```

Comment:
In Ray, this used to look more like

```python
if self.plasma_client.alive:
    libplasma.release(self.plasma_client.conn, self.plasma_id)
```

I think the motivation for this was that Ray (on a single machine) by default kills the plasma store manually at the end of a script. After this, a `PlasmaBuffer` can still go out of scope and try to send a `release` message to the store, and the `check_status` will raise an exception leaving some ugly error or warning messages.

Or is this avoided in some other way?